File: 2/Algo II/Practicos/Mis TADS/Implementaciones/TADBooleano.c
type boolean = nat

fun tru () ret x: nat{
	x = 1;
}

fun Fal () ret x: nat {
	x = 0;
}

fun conjuncion (x: boolean,y: boolean) ret b: boolean {
	if (x == tru() && y == tru()){
		b = tru();
	}
	else {b=fal();}
}

fun disyuncion (x:boolean,y:boolean) ret b: boolean {
	if (x == tru() || y == tru()){
		b = tru();
	}
	else {b=fal();}
}

fun implica (x:boolean,y:boolean) ret b:boolean {
	if (x == tru() && y == fal()) {
		b = fal();
	}
	else {b=tru();}
}

fun eq (x:boolean,y:boolean) ret b:boolean {
	if ((x == tru() && y == tru()) || (x==fal() && y==fal())) {
		b = tru();
	} 
	else { b= fal();}
}

fun not (x:boolean) ret x {
	if (x==tru()) {
		x = fal();
	}
	else { 
		x = tru();
	}
}

fun and (x:boolean, y:boolean) ret b:boolean {
	b = x*y
}

fun or (x:boolean,y:boolean) ret b: boolean {
	b = x + y
}

File: 2/Algo II/Practicos/Mis TADS/Especificaciones/TADBooleano.hs
module TADBooleano where
data Boolean = T | F

conjuncion :: Boolean -> Boolean -> Boolean
conjuncion T x = x
conjuncion F x = F

disyuncion :: Boolean -> Boolean -> Boolean
disyuncion T _ = T
disyuncion F x = x

implica :: Boolean -> Boolean -> Boolean
implica T F = F
implica x y = T

eq :: Boolean -> Boolean -> Boolean
eq T T = T
eq T F = F
eq F T = F
eq F F = F

not :: Boolean -> Boolean
not T = F
not F = T